modulemd/v2/modulemd-profile.c

```diff
@@ -0,0 +1,461 @@
+  /* Read the profile name */
+  YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+  if (event.type != YAML_SCALAR_EVENT)
+    {
+      MMD_YAML_ERROR_EVENT_EXIT (error, event, "Missing profile name");
+    }
+  p = modulemd_profile_new ((const gchar *)event.data.scalar.value);
+  yaml_event_delete (&event);
+
+  /* Read in additional attributes */
+  while (!done)
+    {
+      YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+
+      switch (event.type)
+        {
+        case YAML_MAPPING_START_EVENT: in_map = TRUE; break;
+
+        case YAML_MAPPING_END_EVENT:
+          in_map = FALSE;
+          done = TRUE;
+          break;
+
+        case YAML_SEQUENCE_START_EVENT:
+          if (!in_rpms)
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Unexpected list in profile");
+            }
+          if (!in_map)
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Missing mapping in profile entry");
+            }
+          in_rpms = TRUE;
+          break;
+
+        case YAML_SEQUENCE_END_EVENT:
+          if (!in_rpms)
+            MMD_YAML_ERROR_EVENT_EXIT (error, event, "Unexpected end of list");
+          in_rpms = FALSE;
+          break;
+
+        case YAML_SCALAR_EVENT:
+          if (!in_map)
+            MMD_YAML_ERROR_EVENT_EXIT (
+              error, event, "Missing mapping in profile entry");
+          if (in_rpms)
+            {
+              modulemd_profile_add_rpm (
+                p, (const gchar *)event.data.scalar.value);
+            }
+          else if (g_str_equal (event.data.scalar.value, "rpms"))
+            {
+              in_rpms = TRUE;
+            }
+          else if (g_str_equal (event.data.scalar.value, "description"))
+            {
+              value = modulemd_yaml_parse_string (parser, &nested_error);
+              if (!value)
+                MMD_YAML_ERROR_EVENT_EXIT (
+                  error,
+                  event,
+                  "Failed to parse description in profile: %s",
+                  nested_error->message);
+              modulemd_profile_set_description (p, value);
+              g_clear_pointer (&value, g_free);
+            }
+          else
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Unknown key in profile bbodh");
+            }
+          break;
+
+        default:
+          MMD_YAML_ERROR_EVENT_EXIT (
+            error, event, "Unexpected YAML event in profile");
+          break;
+        }
+
+      yaml_event_delete (&event);
+    }
+  return g_steal_pointer (&p);
+}
+
+
+static gboolean
+modulemd_profile_emit_yaml_rpms (ModulemdProfile *self,
+                                 yaml_emitter_t *emitter,
+                                 GError **error)
+{
+  MODULEMD_INIT_TRACE ();
+  int ret;
+  g_autoptr (GError) nested_error = NULL;
+  MMD_INIT_YAML_EVENT (event);
+  g_auto (GStrv) rpms = modulemd_profile_get_rpms_as_strv (self);
```

Comment:
Similarly to the above parser request, could you add a common emitter routine to take a GHashTable and emit its string keys as a YAML sequence?